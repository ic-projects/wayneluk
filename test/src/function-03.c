// exit code: 0
// description: Should support complex recursive calls (quick sort)

/******** IO macros (DON'T CHANGE THIS SECTION) ********/
#define EOF (-1)

/******** Boolean macros (DON'T CHANGE THIS SECTION) ********/
#define TRUE (1)
#define FALSE (0)

/******** Helper macros ********/
#define MAX_LINE_LENGTH (64)
#define MAX_NUM_LINES (256)

/******** Setup the stack (DON'T CHANGE THIS SECTION) ********/
asm("lui $29,0x2400");

/******** IO function defintions (DON'T CHANGE THIS SECTION) ********/
int getc();
void putc(int);

/******** Helper function definitions ********/
void quick_sort(char *list, int first, int last);
int greater_than(char *a, char *b);
void swap(char *a, char *b);

void print(char *buf);
int input(char *buf, int size);

/******** Main entry point ********/
int entry() {
    char lines[MAX_NUM_LINES * MAX_LINE_LENGTH];
    int line_count;
    for (line_count = 0; line_count < MAX_NUM_LINES; line_count++) {
        char* line = &lines[line_count * MAX_LINE_LENGTH];
        int count = input(line, MAX_LINE_LENGTH);
        if (count == EOF) break;
    }
    quick_sort(lines, 0, line_count - 1);
    for (int i = 0; i < line_count; i ++) {
        print(&lines[i * MAX_LINE_LENGTH]);
        print("\n");
    }
    return 0;
}

/******** Helper functions ********/
void quick_sort(char *list, int first, int last) {
    int i, j, pivot;
    if (first < last) {
        pivot = first;
        i = first;
        j = last;
        while (i < j) {
            while (!greater_than(&list[i * MAX_LINE_LENGTH], &list[pivot * MAX_LINE_LENGTH]) && i < last)
                i++;
            while (greater_than(&list[j * MAX_LINE_LENGTH], &list[pivot * MAX_LINE_LENGTH]))
                j--;
            if (i < j) {
                swap(&list[i * MAX_LINE_LENGTH], &list[j * MAX_LINE_LENGTH]);
            }
        }
        swap(&list[pivot * MAX_LINE_LENGTH], &list[j * MAX_LINE_LENGTH]);
        quick_sort(list, first, j - 1);
        quick_sort(list, j + 1, last);
    }
}

int greater_than(char *a, char *b) {
    for (int count = 0; count < MAX_LINE_LENGTH; count++) {
        if (a[count] == 0) return FALSE;
        if (a[count] > b[count]) return TRUE;
        if (a[count] < b[count]) return FALSE;
    }
    return FALSE;
}

void swap(char *a, char *b) {
    char tmp[MAX_LINE_LENGTH];
    int count;
    for (count = 0; count < MAX_LINE_LENGTH; count++) {
        tmp[count] = a[count];
    }
    for (count = 0; count < MAX_LINE_LENGTH; count++) {
        a[count] = b[count];
    }
    for (count = 0; count < MAX_LINE_LENGTH; count++) {
        b[count] = tmp[count];
    }
}

int input(char *buf, int size) {
    int count;
    for (count = 0; count < size - 1; count++) {
        int c = getc();
        if (c == EOF && count == 0) return -1;
        if (c == EOF || c == '\n') break;
        buf[count] = (char) c;
    }
    buf[count] = 0;
    return count;
}

void print(char *buf) {
    while(*buf) {
        putc(*buf);
        buf++;
    }
}

/******** IO functions (DON'T CHANGE THIS SECTION) ********/
int getc() {
    volatile int *p = (int *) 0x30000000;
    return *p;
}

void putc(int x) {
    volatile int *p = (int *) 0x30000004;
    *p = x;
}
