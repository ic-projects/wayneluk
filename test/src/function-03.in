mmpkyovpjakwtqnvrpsvldnxvayilkuk
waqtjrqmldsznxlzknyzfcgazatfebsk
tfbtxmbsvskpuswzsjadiohwykwezzhe
eqghufvzeilrelljnkyguhojzvclylmz
yomygpavgwmiybhvrtclijtnvskeccdn
yqwvczvxhqlkfhjvwbheexiqgvuifuoi
vrvhzdwupfeegafbanxmwemycerjqqng
ikfsxlvkxekcfabulfkjlkgjsyrxhvke
tdgjmwsecldgcuesogwnaajzkohabzlu
zvrlmlaauxzyyqthpbetggzerqpdyrdx
cpcgcgafkuewgdnoaiigwyekvbfzppoa
ibcmkvaxyuwrqhxvsnudlsmvietswahu
utdpgvpsozjwnmfxmcdezdylikbsfjdq
xqhvakiprqtlayiekgwaljyucsvoiwwm
jzkqlrdxytljntspbriyyqpqvbtmpwwr
qntstoqxbqbcmiretbsmeprcobmgoexq
znxiovfvydxrkhelxauasdzwntzgbtmz
cronxozfuypexctmldjeyywepzhpapft
dgrltzyxntyyroxvcdrcezaxznilwjvd
xdabzpufzkkuqqhilldsueuvkhgwghef
nggxpyyjdofvqmivxxwqohxlerpchxwt
puwsmcttypknvirilnjmddmsfkhpymla
fxjmgpmvozclhflrpbrrnxevkqhbvgim
zovogshnanjwbtghzxnsfvqkpdzrczdy
xbhqhamwenhjmabdvzxfynpxvjcbesgd
dxmstpqbtxlolofizwpcqftgckzkqkfc
wlxmvcjzyfkdyhmnvjevqijcjsholrxy
rnrpcigyqklmdkbhufpgsffkxpdifoax
vjylyhmeumwfiiceyzmlrlkfwtbqflhe
eaargokcxjojopwvydzdiaanwcvgevbm
rnpcpchhuffkabjhieewjvoltxzoykmv
ytzfbyphwmrgatmptbhqnxbnamhxgzyl
dsqfgyqxntzcrlzhqabegfzvtdpcyzsn
fnnrclabytbogzrqkprrppcazqxduyti
pqqnkxqetztyomfelqaykjejpglnygyq
rqxmpdoumgcsatdqrsmzthdcutkcyjse
mqxzgqctgwtqnvvlusqcnsrbkdxnysbx
hythdjqxyxgbeadyavxrehohlkupyjrj
ponvjgqfwjnwqrcizknypmsdvaoqmdfu
uksvilvbsgaqkomqrrpnpmfsfzlyhiwf
fokswkjjcpgzpchjzfmepqzpzgnwullb
rcqgoydloxwlifgnjomukilxmnwiyufw
qlpajkqzatoprlucewkwfzokbxbtxfct
htmcqptzahzveabqzmeksmljczvileew
tvlfxaoazgscjmrggbjzphabxreeqkqp
flnzwggqegmnrmnanoscmohgccivpasd
itbvbtfebxqgpochebzeeirmhhfvthmp
yneapfwjsuaulvldqqljwthfmnngvbso
mygioeevyweibmlrwkipjqrchvyvmszg
kvcdktqsgapvzbxexnghrzfbotumgktp
iatyjuhpwyygjivohfjfnqyilhnkygdr
kgyqqajbcbkyqddfzvvzrajnfvilncmw
xsbmtnvohddzaylrsgfgemrgjtowdtgs
brgpungcagijdhcxkrwqqilfnmtpwqsb
sdtcktpwplebnqowtoludnvazqjbgupj
bbeyjlgpbwclkufjsqtiuumbegehbhen
fiqwawkxeckabvjjxsckclozxyvwkoik
wxztzodrsllfpziozrrauezmawjwqrdf
vvhuwwtdgvfhxmxcizmexsfqdzhwopjw
lbqywyjujnwjzkbdlgyzavyuyvfydgzu
dleaoorncfokcdkpjjcxjvlohchcxasp
pfxktnywugsrzuhuotprlmabepigrnwv
wvywfkdlyrpshrprdzorbmpqlbhyoudb
rrvmylqsvioypycprrqmzaqdjjaoqlnk
dxdfmpedegukzciavzuthrkdtxcjsiis
gsllothggyyusqgdzzaczhaeitmymdxf
ljtgmjdjncwultrglaplegwzleqmcpbf
iaxspzfxeneuusobysmnoqaghnsgywtb
oxouaysripajvyhbgtdnsyewsezyqyfq
vldpdvrkyttduyklfhtifeuwhtmwuhyl
vlvogrcakdlkpwaxdjyhlgjgmotunkhq
catifhujrkdwficbguspdaduakebknfl
iuqnkpxafooyfrjosolafgfdlhqxrkyw
eikfsaigucnruxnbybhnchhfzjexeynn
tcbvdbpvfacsykxswwgpogbnhrpczvlc
umfwgsdkmhlevksxqzanxogipubzorit
utzniaxjmnbcnqzinwxrfjpjphcvxtmm
wlqvnilvhdoxqlzilnisbdzucflfzcvh
bwoljvgjavfrluecujirwgovpzctnmgv
kpfoekqmwxuldzctgizjeufprpsyoddx
yglswgrzjyigifccuksasncvpjbmqdyz
ujcngqcvnxnpcotuyfmgqnurwiivibfl
miawfgumkthismjsrknbbzpvgzacyuok
uxugoobmuejesjpfkdnbujjnywkdkmwu
fqxquyzztqcpckqhkaybwxutyciodusq
bflyfmlqxdwcfspfsceirrojbcwodyat
kqrzabtdxykhdoiesxdpopcwepdssgrg
otgegodtqayimwpwwdepaontvlqpcpgr
jetyxgvgodcigacpzsfisnncjzwcinvt
eemzumddvcrsphxktjdgoqywfucpyjsf
rfinyzwsoisdamfrvgvpjhkqsjdfoafo
gogogiqnqdjzpwfydwbjdyxsrisixvtg
faxxtsxmlzyascupphpiekciivgzvkse
gdcnfuywbsytpuclppxshxvxnjzmtzhf
qwpaaclayexvrgcacgntrowhcymrutjf
snbvhnxeqckidzjeovtxgmsptaorbbon
ahcpjgdaflslogrlrepggeauelegjxqk
szqsoizxybeqopczyhpisqluosugcasm
uubnanxaexjeiszkbhvlybrzqaumonxq
svakyopgmqrqooxefhcrxxduxhzlrgkw
nofumrmqkyczqfvdenykpkihvvjmrbrr
fklevkruxzzhfgylbdjzjiswrzxyudxp
uwhbcyesgcisgosdvyjmfmtzbojexciu
zhhtloglnlpimgellhvclyyfaunkonnm
dybhoyaszqqracyijsivjprmucnojtrz
adzygozzwvfnafovzoakrvsrbnqupsrz
txisocjhigbjipiofvkatmursiiixhwl
rsvvjoaoglkfuwgstnuhncctnepbhceg
iahyktgjacmzadxzcuzjzybycupnlngb
raiiyyagluoyvzbzvfhqjlydoqhatlis
mylstytzxcllluaqpscclbnpikfbcnkt
jwmkqybvljmxxvtfdfngrqnasjtpgnru
klysaytbptccheuvojooyfexwafqkbwb
hzccwpqzmftzjdqmdzlunbfjytdjzlqx
zzvhvfguntcxxehclzgfkvcyflygyayk
jaqqdufrqtnjxudkpydeqjmmqbedsvra
luuuxrehnqaulfbwurdzocfncddagybq
bwvlxsqnibdnlahxsqeokmxyzxztcpdt
iyelvjkpujawbiokfapwyjmalttjneau
xairjujdkucxbcsilyjkatyloqxepxdu
ttsepyxaieznwurqfooqkhpksphkfqki
zgryjokeozrjartbcoqsqpdxesjikfbl
vfwykjjfpwqrjfedqvpeoudncukepjga
daxmgssbpzffdcgxmtepglkgncvwhkfg
upggnfqyvziekmnjgwdodhnudmpjppzr
egqmjmstftsgcaduukqcvnuhjuvrdvad
bvimifhrdpufsyvsuqajyodipcicfwln
exkjwbgoblfxzcxkwoertuqulqnieplc
