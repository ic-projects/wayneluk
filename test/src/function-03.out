adzygozzwvfnafovzoakrvsrbnqupsrz
ahcpjgdaflslogrlrepggeauelegjxqk
bbeyjlgpbwclkufjsqtiuumbegehbhen
bflyfmlqxdwcfspfsceirrojbcwodyat
brgpungcagijdhcxkrwqqilfnmtpwqsb
bvimifhrdpufsyvsuqajyodipcicfwln
bwoljvgjavfrluecujirwgovpzctnmgv
bwvlxsqnibdnlahxsqeokmxyzxztcpdt
catifhujrkdwficbguspdaduakebknfl
cpcgcgafkuewgdnoaiigwyekvbfzppoa
cronxozfuypexctmldjeyywepzhpapft
daxmgssbpzffdcgxmtepglkgncvwhkfg
dgrltzyxntyyroxvcdrcezaxznilwjvd
dleaoorncfokcdkpjjcxjvlohchcxasp
dsqfgyqxntzcrlzhqabegfzvtdpcyzsn
dxdfmpedegukzciavzuthrkdtxcjsiis
dxmstpqbtxlolofizwpcqftgckzkqkfc
dybhoyaszqqracyijsivjprmucnojtrz
eaargokcxjojopwvydzdiaanwcvgevbm
eemzumddvcrsphxktjdgoqywfucpyjsf
egqmjmstftsgcaduukqcvnuhjuvrdvad
eikfsaigucnruxnbybhnchhfzjexeynn
eqghufvzeilrelljnkyguhojzvclylmz
exkjwbgoblfxzcxkwoertuqulqnieplc
faxxtsxmlzyascupphpiekciivgzvkse
fiqwawkxeckabvjjxsckclozxyvwkoik
fklevkruxzzhfgylbdjzjiswrzxyudxp
flnzwggqegmnrmnanoscmohgccivpasd
fnnrclabytbogzrqkprrppcazqxduyti
fokswkjjcpgzpchjzfmepqzpzgnwullb
fqxquyzztqcpckqhkaybwxutyciodusq
fxjmgpmvozclhflrpbrrnxevkqhbvgim
gdcnfuywbsytpuclppxshxvxnjzmtzhf
gogogiqnqdjzpwfydwbjdyxsrisixvtg
gsllothggyyusqgdzzaczhaeitmymdxf
htmcqptzahzveabqzmeksmljczvileew
hythdjqxyxgbeadyavxrehohlkupyjrj
hzccwpqzmftzjdqmdzlunbfjytdjzlqx
iahyktgjacmzadxzcuzjzybycupnlngb
iatyjuhpwyygjivohfjfnqyilhnkygdr
iaxspzfxeneuusobysmnoqaghnsgywtb
ibcmkvaxyuwrqhxvsnudlsmvietswahu
ikfsxlvkxekcfabulfkjlkgjsyrxhvke
itbvbtfebxqgpochebzeeirmhhfvthmp
iuqnkpxafooyfrjosolafgfdlhqxrkyw
iyelvjkpujawbiokfapwyjmalttjneau
jaqqdufrqtnjxudkpydeqjmmqbedsvra
jetyxgvgodcigacpzsfisnncjzwcinvt
jwmkqybvljmxxvtfdfngrqnasjtpgnru
jzkqlrdxytljntspbriyyqpqvbtmpwwr
kgyqqajbcbkyqddfzvvzrajnfvilncmw
klysaytbptccheuvojooyfexwafqkbwb
kpfoekqmwxuldzctgizjeufprpsyoddx
kqrzabtdxykhdoiesxdpopcwepdssgrg
kvcdktqsgapvzbxexnghrzfbotumgktp
lbqywyjujnwjzkbdlgyzavyuyvfydgzu
ljtgmjdjncwultrglaplegwzleqmcpbf
luuuxrehnqaulfbwurdzocfncddagybq
miawfgumkthismjsrknbbzpvgzacyuok
mmpkyovpjakwtqnvrpsvldnxvayilkuk
mqxzgqctgwtqnvvlusqcnsrbkdxnysbx
mygioeevyweibmlrwkipjqrchvyvmszg
mylstytzxcllluaqpscclbnpikfbcnkt
nggxpyyjdofvqmivxxwqohxlerpchxwt
nofumrmqkyczqfvdenykpkihvvjmrbrr
otgegodtqayimwpwwdepaontvlqpcpgr
oxouaysripajvyhbgtdnsyewsezyqyfq
pfxktnywugsrzuhuotprlmabepigrnwv
ponvjgqfwjnwqrcizknypmsdvaoqmdfu
pqqnkxqetztyomfelqaykjejpglnygyq
puwsmcttypknvirilnjmddmsfkhpymla
qlpajkqzatoprlucewkwfzokbxbtxfct
qntstoqxbqbcmiretbsmeprcobmgoexq
qwpaaclayexvrgcacgntrowhcymrutjf
raiiyyagluoyvzbzvfhqjlydoqhatlis
rcqgoydloxwlifgnjomukilxmnwiyufw
rfinyzwsoisdamfrvgvpjhkqsjdfoafo
rnpcpchhuffkabjhieewjvoltxzoykmv
rnrpcigyqklmdkbhufpgsffkxpdifoax
rqxmpdoumgcsatdqrsmzthdcutkcyjse
rrvmylqsvioypycprrqmzaqdjjaoqlnk
rsvvjoaoglkfuwgstnuhncctnepbhceg
sdtcktpwplebnqowtoludnvazqjbgupj
snbvhnxeqckidzjeovtxgmsptaorbbon
svakyopgmqrqooxefhcrxxduxhzlrgkw
szqsoizxybeqopczyhpisqluosugcasm
tcbvdbpvfacsykxswwgpogbnhrpczvlc
tdgjmwsecldgcuesogwnaajzkohabzlu
tfbtxmbsvskpuswzsjadiohwykwezzhe
ttsepyxaieznwurqfooqkhpksphkfqki
tvlfxaoazgscjmrggbjzphabxreeqkqp
txisocjhigbjipiofvkatmursiiixhwl
ujcngqcvnxnpcotuyfmgqnurwiivibfl
uksvilvbsgaqkomqrrpnpmfsfzlyhiwf
umfwgsdkmhlevksxqzanxogipubzorit
upggnfqyvziekmnjgwdodhnudmpjppzr
utdpgvpsozjwnmfxmcdezdylikbsfjdq
utzniaxjmnbcnqzinwxrfjpjphcvxtmm
uubnanxaexjeiszkbhvlybrzqaumonxq
uwhbcyesgcisgosdvyjmfmtzbojexciu
uxugoobmuejesjpfkdnbujjnywkdkmwu
vfwykjjfpwqrjfedqvpeoudncukepjga
vjylyhmeumwfiiceyzmlrlkfwtbqflhe
vldpdvrkyttduyklfhtifeuwhtmwuhyl
vlvogrcakdlkpwaxdjyhlgjgmotunkhq
vrvhzdwupfeegafbanxmwemycerjqqng
vvhuwwtdgvfhxmxcizmexsfqdzhwopjw
waqtjrqmldsznxlzknyzfcgazatfebsk
wlqvnilvhdoxqlzilnisbdzucflfzcvh
wlxmvcjzyfkdyhmnvjevqijcjsholrxy
wvywfkdlyrpshrprdzorbmpqlbhyoudb
wxztzodrsllfpziozrrauezmawjwqrdf
xairjujdkucxbcsilyjkatyloqxepxdu
xbhqhamwenhjmabdvzxfynpxvjcbesgd
xdabzpufzkkuqqhilldsueuvkhgwghef
xqhvakiprqtlayiekgwaljyucsvoiwwm
xsbmtnvohddzaylrsgfgemrgjtowdtgs
yglswgrzjyigifccuksasncvpjbmqdyz
yneapfwjsuaulvldqqljwthfmnngvbso
yomygpavgwmiybhvrtclijtnvskeccdn
yqwvczvxhqlkfhjvwbheexiqgvuifuoi
ytzfbyphwmrgatmptbhqnxbnamhxgzyl
zgryjokeozrjartbcoqsqpdxesjikfbl
zhhtloglnlpimgellhvclyyfaunkonnm
znxiovfvydxrkhelxauasdzwntzgbtmz
zovogshnanjwbtghzxnsfvqkpdzrczdy
zvrlmlaauxzyyqthpbetggzerqpdyrdx
zzvhvfguntcxxehclzgfkvcyflygyayk
